File: python/BSpline.py
import numpy as np
import matplotlib.pyplot as plt
import sys

class BSpline:

    def __init__(self, samples, knot_vector, order=3, precision=1e-8, clampRight=False, clamp=False):
        
        self.samples = samples
        self.knot_vector = list(knot_vector)
        self.order = order
        self.precision = precision

        self.clamp = clamp
        if clamp:
            self.og_knot_vector	= knot_vector
            self.__clamp_ends()

        self.clampRight = clampRight 
        if clampRight and not clamp:
            self.og_knot_vector = knot_vector
            self.__clamp_right()
        
        self.spline = self.__spline_function()

    def __clamp_ends(self):
        if self.clamp:
            for i in range(self.order):
                self.knot_vector.insert(0, self.knot_vector[0])
                self.knot_vector.insert(-1, self.knot_vector[-1])

    def __clamp_right(self):
        if self.clampRight:
            for i in range(self.order):
                self.knot_vector.insert(-1, self.knot_vector[-1])
                
    def __find_interval(self, value):

        knot_vector = self.knot_vector
        n_knots = len(knot_vector)
    
        if value < knot_vector[0] or value > knot_vector[-1]:
            sys.exit("Value must be within the domain of the knot vector!")

        for i in range(n_knots-1):
            if value >= knot_vector[i] and value < knot_vector[i+1]:
                return i

    def __get_basis_value(self, value):

        order = self.order
        knot_vector = self.knot_vector
        n_knots = len(knot_vector)
        
        if order > n_knots-2:
            order = n_knots-2
            print(f'Warning: Order reduced to maximum possible!')
        if order < 0:
            sys.exit("The polynomial order has to be a positive integer!")

        def omega(x, ti, tik):
            if ti == tik:
                return 0.
            else:
                return (x-ti)/(tik-ti)
            
        basis = np.zeros((n_knots-1,n_knots-1))
        basis[0][self.__find_interval(value)] = 1

        for k in range(1, order+1):
            N = basis[k-1]

            for i in range(n_knots-k-1):
                basis[k][i] = omega(value, knot_vector[i], knot_vector[i+k])*N[i] + (1. - omega(value, knot_vector[i+1], knot_vector[i+k+1]))*N[i+1]
                    
        return basis[order][0:n_knots-order-1]
    
    def __get_collocation_matrix(self, size=100, precision=1e-8):

        order = self.order
        n_knots = len(self.knot_vector)
        domain = np.linspace(self.knot_vector[0], self.knot_vector[-1]-precision, size)
        collocationMatrix = np.zeros((n_knots-order-1, size))

        for i,x in enumerate(domain):
            collocationMatrix[:,i] = self.__get_basis_value(x)

        return collocationMatrix

    def __fit_spline(self, collocationMatrix):

        aTa = np.matmul(collocationMatrix, collocationMatrix.T)
        aTa_inv = np.linalg.inv(aTa)
        aTa_inv_aT = np.matmul(aTa_inv, collocationMatrix)
        fit_coeff = np.matmul(aTa_inv_aT, self.samples)

        return fit_coeff

    def __spline_function(self):

        collocationMatrix = self.__get_collocation_matrix(size=len(self.samples), precision=self.precision)

        #self.plot_basis('test_basis')
        self.fit_coeff = self.__fit_spline(collocationMatrix)

        #construct spline function using fit coefficients
        spline = [np.matmul(self.fit_coeff, basis_slice) for basis_slice in collocationMatrix.T]

        #calculate error
        diff = np.subtract(spline, self.samples)
        self.error = 0.5*np.matmul(diff, diff)
        
        return spline

    def get_collocation_matrix(self, size=100, precision=1e-8):
        return self.__get_collocation_matrix(size=size, precision=precision)
    
    def spline_function(self):
        return self.spline

    def get_fit_coeff(self):
        return self.fit_coeff

    def get_error(self):
        #print(f'knot vector: {self.knot_vector}, error: {self.error}')
        return self.error

    def plot_fit(self, name):
        from scipy import interpolate

        knot_vector = self.knot_vector
        size = len(self.samples)
        domain = np.linspace(knot_vector[0], knot_vector[-1], size)
        interpolated_function = interpolate.interp1d(x=domain, y=self.samples, kind='cubic')
        interpolated_waveform_at_knots = interpolated_function(knot_vector)
        
        plt.xlabel('Domain')
        plt.ylabel('Amplitude')
        plt.plot(knot_vector, interpolated_waveform_at_knots, 'o', color='black', label='Interpolation at Knots')
        plt.plot(domain, self.samples, label='Interpolated data')
        plt.plot(domain, self.spline, label='B-Spline Fit')
        plt.grid()
        plt.text(0.75,0.25,f'Error: {round(self.error,5)}')
        legend=plt.legend(loc="upper right", fancybox=True)
        frame = legend.get_frame()
        frame.set_edgecolor('black')
        frame.set_alpha(1)
        plt.savefig(name+'.pdf')
        print(f'Saved plot: {name}.pdf')
        plt.clf()
    
    def plot_basis(self, name, size=100, precision=1e-8):

        order = self.order
        knot_vector = self.knot_vector
        n_knots = len(knot_vector)
        domain = np.linspace(knot_vector[0], knot_vector[-1]-precision, size)
        collocationMatrix = self.__get_collocation_matrix(size)

        for idx, arr in enumerate(collocationMatrix):
            plt.plot(domain, arr, label=f'B({idx})')

        y_knots = np.zeros(n_knots)
        plt.plot(knot_vector, y_knots, 'o', color='black')

        plt.xlabel('Domain')
        plt.ylabel('Spline Basis Value')
        plt.grid()
        legend=plt.legend(loc="upper right", title=f'k={order}', fontsize='small', fancybox=True)
        frame = legend.get_frame()
        frame.set_edgecolor('black')
        frame.set_alpha(1)
        plt.savefig(f'{name}_basis_order{order}_size{int(size)}.pdf')
        print(f'Saved plot: {name}_basis_order{order}_size{int(size)}.pdf')
        plt.clf()
        

File: src/BSpline.cc
#include "BSpline.h"

using namespace arma;

BSpline::BSpline(const std::vector<double> &knot_vector, const int order, const double precision):
  knot_vector_(knot_vector),
  n_knots_(int(knot_vector.size())),
  precision_(precision),
  order_(order),
  ls_error_(-999.){

  if(order_ > n_knots_-2 || order_ < 0)
    throw std::invalid_argument("Polynomial order has to be a positive integer no larger than the number of knots minus 2!");
  
}

BSpline::~BSpline() = default;

//public
double BSpline::GetError() const{
  return ls_error_;
}

std::vector<double> BSpline::SplineLS(const std::vector<double> &samples){

  int sample_size = int(samples.size());

  dmat collocation_matrix = CollocationMatrix(sample_size);
  dvec fit_coeff = CoeffLeastSquare(samples, collocation_matrix);

  dvec spline;
  SplineFunction(spline, samples, fit_coeff, collocation_matrix);

  return {spline.begin(), spline.end()};
}

//private
int BSpline::FindInterval(const double value) const{

  if(value < knot_vector_[0] || value > knot_vector_[n_knots_-1])
    throw std::invalid_argument("Value must be within the domain of the knot vector!");

  int interval = -1;
  
  for(int i = 0; i < n_knots_-1; i++)
    if (value >= knot_vector_[i] && value < knot_vector_[i+1])
      interval = i;

  return interval;
}

dvec BSpline::BasisValuesAt(const double t) const{

  auto omega = [](double x, double ti, double tik){
    if(ti == tik)
      return 0.;
    else
      return (x-ti)/(tik-ti);
  };

  dmat basis = zeros<dmat>(n_knots_-1, n_knots_-1); 
  basis(0, FindInterval(t)) = 1.;

  for(uword k = 1; k < order_+1; k++){
    drowvec B = basis.row(k-1);
    for(uword i = 0; i < n_knots_-k-1; i++)
      basis(k,i) = omega(t, knot_vector_[i], knot_vector_[i+k])*B(i) + (1. - omega(t, knot_vector_[i+1], knot_vector_[i+k+1]))*B(i+1);
  }

  dvec basis_slice = basis(order_, span(0, n_knots_-order_-2)).t();

  return basis_slice;
}

dmat BSpline::CollocationMatrix(const int size) const{

  vec domain = linspace(knot_vector_[0], knot_vector_[n_knots_-1]-precision_, size);

  dmat collocation_matrix = zeros<dmat>(n_knots_-order_-1, size);
  
  for(uword i = 0; i < domain.size(); i++)
    collocation_matrix.col(i) = BasisValuesAt(domain(i));

  return collocation_matrix;
}

dvec BSpline::CoeffLeastSquare(const dvec &samples, const dmat &collocation_matrix) {
  
  //return (inv_sympd(collocation_matrix*collocation_matrix.t())*collocation_matrix)*samples;
  return solve(collocation_matrix*collocation_matrix.t(), collocation_matrix)*samples;
}

dvec& BSpline::SplineFunction(dvec &spline, const dvec &samples, const dvec &coeff, const dmat &collocation_matrix){

  int sample_size = int(samples.size());
  spline = zeros<dvec>(sample_size);

  for(int i = 0; i < sample_size; i++)
    spline(i) = dot(coeff, collocation_matrix.col(i));

  dvec diff = spline-conv_to<dvec>::from(samples);
  ls_error_ = dot(diff, diff);
  
  return spline;
}


void PlotSplineFit(const std::vector<double> &waveform,
		   const std::vector<double> &spline,
		   const std::vector<double> &knot_vector,
		   const std::string &name){

  std::vector<int> colors = {kBlack, kOrange+1};

  // Create TGraph objects for each data set
  const int data_size = int(waveform.size());
  std::vector<double> x_data = LinSpaceVec(knot_vector.front(), knot_vector.back(), data_size);
  std::vector<double> knot_values(knot_vector.size(), 0.);
  ShannonNyquistInterp(x_data, waveform, knot_vector, knot_values);

  std::vector<TGraph*> graphs(3, nullptr);
  graphs[0] = new TGraph(data_size, &x_data[0], &waveform[0]); 
  graphs[1] = new TGraph(data_size, &x_data[0], &spline[0]);
  auto g_knots = new TGraph(int(knot_vector.size()), &knot_vector[0], &knot_values[0]);
  
  // Create a TMultiGraph object to hold the TGraphs
  auto multi_graph = new TMultiGraph();
  for(int i = 0; i < graphs.size()-1; i++){
    graphs[i]->SetLineWidth(3);
    graphs[i]->SetLineColor(short(colors[i]));
    graphs[i]->SetMarkerSize(0);
    multi_graph->Add(graphs[i]);
  }

  g_knots->SetMarkerStyle(20);
  g_knots->SetMarkerColor(kBlack);
  g_knots->SetLineWidth(0);
  //multi_graph->Add(graphs[2]);

  // Create a canvas to draw the TMultiGraph
  auto canvas = new TCanvas("canvas", "canvas");

  // create the legend
  auto legend = new TLegend(0.6, 0.6, 0.8, 0.8);
  legend->AddEntry(graphs[0], "Interpolated Data", "l");
  legend->AddEntry(graphs[1], "Spline", "l");
  legend->AddEntry(g_knots, "Knots", "p");
  legend->SetBorderSize(0);
  legend->SetFillStyle(0);
  
  // Draw the TMultiGraph with proper axis labels and title
  multi_graph->SetTitle("");
  multi_graph->GetXaxis()->SetTitle("Domain");
  multi_graph->GetXaxis()->CenterTitle();
  multi_graph->GetYaxis()->SetTitle("Amplitude");
  multi_graph->GetYaxis()->CenterTitle();
  multi_graph->Draw("AL");
  g_knots->Draw("Psame");

  legend->Draw("same");
  canvas->Update();
  // Save the canvas to a PDF file with the given name
  std::string filename = name + ".pdf";
  canvas->SetGrid();
  canvas->SaveAs(filename.c_str());
  
  // Clean up the TGraphs, TMultiGraph, and TCanvas
  for (TGraph* graph : graphs) {
    delete graph;
  }
  delete g_knots;
  delete multi_graph;
  delete canvas;
}
